lib/elixir/lib/module/types/descr.ex

```diff
@@ -495,4 +525,539 @@ defmodule Module.Types.Descr do
       _ -> nil
     end
   end
+
+  ## Not_set
+
+  # `not_set()` is a special base type that represents an unset field in a map.
+  # E.g., `%{a: integer(), b: not_set(), ...}` represents a map with an integer
+  # field `a` and an unset field `b`, and possibly other fields.
+
+  # The `if_set()` modifier is syntactic sugar for specifying the key as a union
+  # of the key type and `not_set()`. For example, `%{:foo => if_set(integer())}`
+  # is equivalent to `%{:foo => integer() or not_set()}`.
+
+  # `not_set()` has no meaning outside of map types.
+
+  def not_set(), do: @unset
+  def if_set(type), do: Map.update(type, :bitmap, @bit_unset, &(&1 ||| @bit_unset))
+
+  defp has_unset?(type), do: (Map.get(type, :bitmap, 0) &&& @bit_unset) != 0
+
+  defp remove_not_set(type) do
+    case type do
+      %{:bitmap => @bit_unset} -> Map.delete(type, :bitmap)
+      %{:bitmap => bitmap} -> Map.put(type, :bitmap, bitmap &&& ~~~@bit_unset)
+      _ -> type
+    end
+  end
+
+  ## Map
+  #
+  # Map operations.
+  #
+  # Maps are in disjunctive normal form (DNF), that is, a list (union) of pairs
+  # `{map_literal, negated_literals}` where `map_literal` is a map type literal
+  # and `negated_literals` is a list of map type literals that are negated from it.
+  #
+  # A map literal is a pair `{:open | :closed, %{keys => value_types}}`.
+  #
+  # For instance, the type `%{..., a: integer()} and not %{b: atom()}` can be represented
+  # by the DNF containing one pair, where the positive literal is `{:open, %{a => integer()}}`
+  # and the negated literal is `{:closed, %{b => atom()}}`.
+  #
+  # The goal of keeping symbolic negations is to avoid distributing difference on
+  # every member of a union which creates a lot of map literals in the union and
+  # requires emptiness checks to avoid creating empty maps.
+  #
+  # For instance, the difference between `%{...}` and `%{a: atom(), b: integer()}`
+  # is the union of `%{..., a: atom(), b: if_set(not integer())}` and
+  # `%{..., a: if_set(not atom()), b: integer()}`; for maps with more keys,
+  # each key in a negated literal may create a new union when eliminated.
+  #
+  # Set-theoretic operations take two DNFs (lists) and return a DNF (list).
+  # Simplifications can be done to prune the latter.
+
+  # Create a DNF from a specification of a map type.
+  defp map_new(open_or_closed, pairs) do
+    fields =
+      Map.new(pairs, fn
+        {{:optional, [], [key]}, type} ->
+          {key, {true, type}}
+
+        {key, type} ->
+          if has_unset?(type),
+            do: {key, {true, remove_not_set(type)}},
+            else: {key, {false, type}}
+      end)
```

Comment:
```suggestion
      Map.new(pairs)
```

PS: we don't need to have tuples as values.